doc/specs/stdlib_string_type.md

```diff
@@ -0,0 +1,1607 @@
+---
+title: string type
+---
+
+# The `stdlib_string_type` module
+
+[TOC]
+
+## Introduction
+
+The `stdlib_string_type` provides a derived type holding an arbitrary sequence
+of characters compatible with most Fortran intrinsic character procedures as
+well as operators for working with character variables and constants.
+
+
+## Derived types provided
+
+
+<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
+### The `string_type` derived type
+
+The `string_type` is defined as a non-extenable derived type representing a
+sequence of characters. The internal representation of the character sequence
+is implementation dependent and not visible for the user of the module.
+
+#### Status
+
+Experimental
+
+
+## Procedures and methods provided
+
+Procedures returning `string_type` instances can usually used in elemental
+context, while procedures returning scalar character values can only be
+used in a pure way.
+
+
+<!-- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -->
+### Constructor for empty string
+
+The module defines a default constructor to create an empty string type.
+
+#### Description
+
+Creates a string instance representing an empty string.
+
+#### Syntax
+
+`res = [[stdlib_string_type(module):string_type(interface)]] ()`
+
+#### Status
+
+Experimental
+
+#### Class
+
+Elemental function.
+
+#### Argument
+
+None.
+
+#### Result value
+
+The result is an instance of `string_type`.
+
+#### Example
+
+```fortran
+use stdlib_string_type
+implicit none
+type(string_type) :: string
+string = string_type()
+! len(string) == 0
+end
```

Comment:
```suggestion
program demo_string_type
    use stdlib_string_type
    implicit none
    type(string_type) :: string
   string = string_type()
   ! len(string) == 0
end program demo_string_type
```

Reply:
Personally, I like the short example without the explicit `program` block. Is this worthy of a documentation meta-issue?

I might even feel tempted to drop the `implicit none` for brevity, although I imagine this is a controversial stance. :see_no_evil:

Reply:
I left out the program block to help focus on the short example rather than on boilerplate. I can add program blocks to make it more clear that we have standalone programs here.

Regarding `implicit none`, good practice wants to be enforced and we don't want to start another lengthy thread on this old topic *again*.

In any case, let's discuss this elsewhere, GitHub PRs get clutter quickly once a discussion starts.

Reply:
Sorry for forgetting the  `implicit none`. It should be in the suggestion. I wil add it ;)

Reply:
> Personally, I like the short example without the explicit `program` block. Is this worthy of a documentation meta-issue?

I believe it has been discussed somewhere when opening one of the first PR (I cannot retrieve it). The idea was to have some tools that would extract them, compile them and test them.

IMO adding program blocks doesn't add complexity (it is only one additional line ). However, my main issue with this short form is that it might not be obvious for Fortran beginners.